ambry-router/src/test/java/com/github/ambry/router/NonBlockingRouterTest.java

```diff
@@ -1367,6 +1365,258 @@ public void onEvent(ReplicaId replicaId, Object e) {
+    routerMetadataCacheErrorOnGetCompositeBlob(ServerErrorCode.Blob_Deleted, RouterErrorCode.BlobDeleted, true);
+  }
+  @Test
+  public void testRouterMetadataCacheBlobExpiredOnGetCompositeBlob() throws Exception {
+    routerMetadataCacheErrorOnGetCompositeBlob(ServerErrorCode.Blob_Expired, RouterErrorCode.BlobExpired, true);
+  }
+  @Test
+  public void testRouterMetadataCacheUnexpectedInternalErrorOnGetCompositeBlob() throws Exception {
+    routerMetadataCacheErrorOnGetCompositeBlob(ServerErrorCode.Unknown_Error, RouterErrorCode.UnexpectedInternalError, false);
+  }
+
+  void routerMetadataCacheErrorOnDeleteCompositeBlob(ServerErrorCode serverErrorCode, RouterErrorCode routerErrorCode) throws Exception {
+    try{
+      Properties properties = getNonBlockingRouterProperties("DC1");
+      properties.setProperty("router.blob.metadata.cache.enabled", Boolean.toString(true));
+      properties.setProperty("router.smallest.blob.for.metadata.cache", Long.toString(0));
+      setRouterWithMetadataCache(properties, new AmbryCacheMetrics(1,1,1,1));
+      AmbryCacheWithMetrics ambryCacheWithMetrics = (AmbryCacheWithMetrics) router.getBlobMetadataCache();
+      assertNotNull("Metadata cache must be created", ambryCacheWithMetrics);
+
+      // put a composite blob and test it is absent in the metadata cache
+      setOperationParams(2 * PUT_CONTENT_SIZE, TTL_SECS);
+      String blobId =
+          router.putBlob(putBlobProperties, putUserMetadata, putChannel, new PutBlobOptionsBuilder().build())
+              .get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+      assertTrue("BlobID must be version 6 blobID", BlobId.getVersion(blobId) == BlobId.BLOB_ID_V6);
+      assertTrue("BlobID must be of type metadata", BlobId.getBlobDataType(blobId) == BlobId.BlobDataType.METADATA);
+
+      // successful get
+      ByteRange range = ByteRanges.fromOffsetRange(1, 2);
+      router.getBlob(blobId, new GetBlobOptionsBuilder().range(range).build()).get(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+      assertTrue("Must encounter a cache miss", ambryCacheWithMetrics.getCacheMissCountDown().await(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+      assertTrue("Must put metadata in cache", ambryCacheWithMetrics.getPutObjectCountDown().await(AWAIT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+      assertNotNull("Blob metadata must be present in metadata cache", router.getBlobMetadataCache().getObject(blobId));
+
+      // failed delete
+      mockServerLayout.getMockServers().forEach(mockServer -> mockServer.setServerErrorForAllRequests(serverErrorCode));
+      TestCallback<Void> testCallback = new TestCallback<>();
+      Future<Void> future = router.deleteBlob(blobId, null, testCallback, null);
+      assertFailureAndCheckErrorCode(future, testCallback, routerErrorCode);
+      assertNotNull("Blob metadata must be present in metadata cache", router.getBlobMetadataCache().getObject(blobId));
```

Comment:
This is not related to tests. But just wanted to note that similar to how we are invalidating cache when blob is not found during GetBlob operations, we should update logic sometime later to invalidate cache when blob is not found during GETBlobInfo, TTLUpdate or DELETE operations too. 

This will avoid BlobInfo call during later GET operations. Just wondering that even though one BlobInfo is not costly, since we already saw the blob is gone, we  should probably make use of it.

Reply:
got it